### PR TITLE
Added status field selector to allocation

### PR DESCRIPTION
**Pull request by jshaofuturewei**
Customer Resource does not support status field selector used by informers. 

This pull request is to add status field selector to allocation crd so that informers can add filter selectors.

1. Run "hack/globalscheduler/globalscheduler-up.sh"
2. Run "cluster/kubectl.sh apply -f globalscheduler/test/yaml/allocation_a1.yaml" to add an allocation
3. Run "cluster/kubectl.sh get allocations --field-selector status.phase=Scheduled" and the new allocation is found
4. Run "cluster/kubectl.sh get allocations --field-selector status.phase=Bound" and there is no resource
5. Run "cluster/kubectl.sh get allocations" and the new allocation is found
6. Run "cluster/kubectl.sh get allocations --field-selector status.distributor_name=b" and the new allocation is found
7. Run "cluster/kubectl.sh get allocations --field-selector status.distributor_name=a" and there is no resource
8.  Run "cluster/kubectl.sh get allocations --field-selector status.scheduler_name=a" and the new allocation is found
9. . Run "cluster/kubectl.sh get allocations --field-selector status.scheduler_name=b" and there is no resource
10. Run "cluster/kubectl.sh get allocations --field-selector status.dispatcher_name=c" and the new allocation is found
11. Run "cluster/kubectl.sh get allocations --field-selector status.dispatcher_name=b" and there is no resource
12.  Run "cluster/kubectl.sh get allocations --field-selector status.cluster_name=d" and the new allocation is found
13.  Run "cluster/kubectl.sh get allocations --field-selector status.cluster_name=f" and there is no resource

### CODE REVIEW

**Review by kimeunju108 — COMMENTED** [review]
I am just wondering why we need this allocation crd. Can we update and use Status in each controller CRD?

**Comment by jshaofuturewei** [comment]
> I am just wondering why we need this allocation crd. Can we update and use Status in each controller CRD?

We discussed it in the meetings with Cathy. Allocation is a "Batch Pod" which have several resource requests in it. The status comes with allocation not controllers.

**Review by CoderKevinZhang — APPROVED** [review]

File: globalscheduler/test/yaml/allocation_a1.yaml [hunk 1]
```diff
@@ -7,31 +7,35 @@ spec:
     name: v1rg
```
Comment (pdgetrf) [inline globalscheduler/test/yaml/allocation_a1.yaml:7]: what does a1 mean in the file name?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go [hunk 2]
```diff
@@ -124,8 +124,10 @@ type crConverter struct {
 
 func (c *crConverter) ConvertFieldLabel(gvk schema.GroupVersionKind, label, value string) (string, string, error) {
 	// We currently only support metadata.hashkey, metadata.ownerReferences.*(Kind), metadata.namespace and metadata.name.
+	// Added status.phase, status.distributorName, status.schedulerName, status.dispatcherName, and status.clusterName
 	switch {
-	case label == "metadata.name" || label == "metadata.hashkey":
+	case label == "metadata.name" || label == "metadata.hashkey" || label == "status.phase" || label == "status.distributor_name" ||
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:129]: for such a long list, might be suitable to put them in a separate function and have a test for it.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 3]
```diff
@@ -189,7 +190,33 @@ func (a customResourceStrategy) GetAttrs(obj runtime.Object) (labels.Set, fields
...
-	return labels.Set(accessor.GetLabels()), objectMetaFieldsSet(accessor, a.namespaceScoped, a.tenantScoped), nil
+	crdObj := obj.(*unstructured.Unstructured)
+	crdRes := crdObj.UnstructuredContent()
+	fieldSet := objectMetaFieldsSet(accessor, a.namespaceScoped, a.tenantScoped)
+	if statusVal, ok := crdRes["status"]; ok {
+		if statusRes, ok := statusVal.(map[string]interface{}); ok {
+			if phase, ok := statusRes["phase"]; ok {
+				fieldSet["status.phase"] = fmt.Sprintf("%v", phase)
+			}
+			if distributorname, ok := statusRes["distributor_name"]; ok {
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:201]: do all these strings like "distributor_name", "scheduler_name" have any corresponding variables?

File: globalscheduler/cmd/proxy-server/allocation_handler.go [hunk 4]
```diff
@@ -168,7 +168,7 @@ func (handler *AllocationHandler) deleteAllocation(w http.ResponseWriter, r *htt
 
 func (handler *AllocationHandler) watchAllocationStatus(namespace, name string, alloc *v1.Allocation, ctx context.Context, watcher watch.AggregatedWatchInterface, timer *time.Timer) error {
```
Comment (pdgetrf) [inline globalscheduler/cmd/proxy-server/allocation_handler.go:169]: watchAllocationStatus --> watchAllocationPhase?

**Review by pdgetrf — APPROVED** [review]
some comments. looks good over all.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go (hunk 3 above)
Comment (jshaofuturewei) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:201]: They will have variables once the application is processed by distributor, scheduler, or dispatcher

File: globalscheduler/cmd/proxy-server/allocation_handler.go (hunk 4 above)
Comment (jshaofuturewei) [reply globalscheduler/cmd/proxy-server/allocation_handler.go:169]: Updated

File: globalscheduler/test/yaml/allocation_a1.yaml (hunk 1 above)
Comment (jshaofuturewei) [reply globalscheduler/test/yaml/allocation_a1.yaml:7]: a1 is the allocation name

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go (hunk 2 above)
Comment (jshaofuturewei) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:129]: It is a common converter for crds. For each crds, it does not have any separator functions to handle selectors.

**Pull request merged**